src/librustc/traits/object_safety.rs

```diff
@@ -163,14 +188,16 @@ impl<'tcx> TyCtxt<'tcx> {
     fn predicates_reference_self(
         self,
         trait_def_id: DefId,
-        supertraits_only: bool) -> bool
-    {
+        self_ty: Ty<'tcx>,
+        supertraits_only: bool,
+    ) -> bool {
         let trait_ref = ty::Binder::dummy(ty::TraitRef::identity(self, trait_def_id));
         let predicates = if supertraits_only {
             self.super_predicates_of(trait_def_id)
         } else {
             self.predicates_of(trait_def_id)
         };
+        let has_self_ty = |t: Ty<'tcx>| t.walk().any(|t| t == self_ty);
```

Comment:
I see some caching was removed (`TypeFlags::HAS_SELF`), but I'm not sure how often this operation is used.
It would be nice to pass this through a perf run.